Added a test for <a download> error case

phistuck opened the pull request:
According to the specification, <a download> should download (unless a user preference indicated otherwise), regardless of the fetching result (404, 200...) and not navigate.
This test verifies that navigation does not occur. It relies on a timeout, which might not be ideal, but I am not sure how else to decide that a navigation did not happen.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241757754)

# Firefox (nightly) #
Testing web-platform-tests at revision 2dc6879c1584ab46a90064cd2aa188b3df15db81
Using browser at version BuildID 20170611100318; SourceStamp 799d43edb324395cf02da6b028e803712334615f
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6211/html/semantics/text-level-semantics/the-a-element/a-download-click-404.html">/html/semantics/text-level-semantics/the-a-element/a-download-click-404.html</a></summary>


|                      Subtest                      | Results | Messages |
|---------------------------------------------------|---------|----------|
|                                                   | OK      |          |
| `Do not navigate to 404 for anchor with download` | PASS    |          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241757755)

# Sauce (safari) #
Testing web-platform-tests at revision 2dc6879c1584ab46a90064cd2aa188b3df15db81
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6211/html/semantics/text-level-semantics/the-a-element/a-download-click-404.html">/html/semantics/text-level-semantics/the-a-element/a-download-click-404.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | TIMEOUT |          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241757761)

# Chrome (unstable) #
Testing web-platform-tests at revision 2dc6879c1584ab46a90064cd2aa188b3df15db81

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/241757762)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 2dc6879c1584ab46a90064cd2aa188b3df15db81

sideshowbarker commented:
w3c-test:mirror

phistuck commented:
@sideshowbarker - I am not sure what your comment is supposed to mean...

Anyway, looks like the test does not really run correctly on Firefox and I am not sure it runs as expected anywhere. I think I will just give up on this one and close it.

sideshowbarker commented:
> @sideshowbarker - I am not sure what your comment is supposed to mean...

it’s a command that gets the PR branch mirrored to https://w3c-test.org/submissions/

phistuck commented:
I see - thank you, that is helpful. :)
http://w3c-test.org/submissions/6211/html/semantics/text-level-semantics/the-a-element/a-download-click-404.html

phistuck committed:
Added a test for <a download> not navigating on 404

phistuck commented:
I think this is ready for a code review.
The test is expected to pass on every browser except Firefox, which navigates on a 404 text/html response.
However, Sauce Labs use desktop Safari 10 and `<a download>` is supported on desktop Safari 10.1 or later, so Safari fails the test.

Note - the Chrome Travis CI failure looks temporary, because it succeeded before I squashed the commits.

sideshowbarker commented:
> I think this is ready for a code review.

It may be a while before anybody can get around to reviewing it, unless you can manage to identify a good reviewer on your own and ask them yourself to review the patch. If it’s someone who hasn’t reviewed here before, we can set them up with read-level access to the repo—which will enable them to be assigned the review—and then once they’ve reviewed it and you’ve responded to the review comments, you can let us know it’s been reviewed and you need somebody to merge it.

zcorpan commented:
@bobholt @jgraham any idea what went wrong in https://travis-ci.org/w3c/web-platform-tests/jobs/242912522 ?

> WebDriverException: Message: unknown error: Chrome failed to start: crashed

(I've now restarted the job.)

zcorpan commented on html/semantics/text-level-semantics/the-a-element/a-download-click-404.html:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
+<meta charset="utf-8">
+<title>Clicking on an &lt;a> element with a download attribute and href that leads to 404 should not navigate</title>
+<link rel="author" title="phistuck" href="mailto:user@domain.com">
```
Is this your actual email? (An author `link` is not required; that info is in the git log anyway.)

phistuck replied:
No. I was wondering whether it is required. I will remove it.

zcorpan commented on html/semantics/text-level-semantics/the-a-element/a-download-click-404.html:
```diff
@@ -0,0 +1,24 @@
+"use strict";
+async_test(function (t) {
+    const errorFrame = document.querySelector("#error-frame");
+    errorFrame.addEventListener("load", function () {
```
This is racy; the load event could fire before this script executes. Use `onload` event handler attribute in the markup, or create and insert the iframe in the script.

phistuck replied:
I was deliberately trying to avoid inline event handlers...
I guess I will create the iFrame within the script.

phistuck replied:
Thank you for the review, I forgot to write before. :)

Are you fine with timeout part, though?
Or since this is expected to run mostly locally, then it is not a problem in this case?
Should I make it shorter?

zcorpan replied:
I think 1s is reasonable here. It would only be unstable on devices where latency is unpredictable *and* there is a bug in behavior.

zcorpan commented:
The test generally LGTM with respect to the current spec. Can you also file an issue for Gecko? And maybe also for WebKit; Safari TP appears to download the file without saying anything about there being an error.

phistuck commented:
@zcorpan - in my (personal) opinion, I think the Chrome behavior is wrong. I would expect the file to download regardless of the HTTP status code or content.

phistuck committed:
Made the test less racy

Reduces the races by constructing the iFrame from within the script and listening to the "load" event earlier.
Also removed non-required meta-data and switched to arrow functions.

phistuck committed:
Fixed a not-yet-constructed <body> race.

phistuck committed:
Fixed a lint error.

Removed extraneous trailing spaces.

phistuck commented:
I filed a Firefox bug - https://bugzilla.mozilla.org/show_bug.cgi?id=1373620

phistuck commented:
@zcorpan - thank you for the review!

How does it get merged?
Do you want me to squash the commits first?

zcorpan commented:
Merged, I just wanted to restart the Chrome job to see if it still errored (which it did 🙁). That seems to be a general problem and not caused by this test.

phistuck commented:
Yeah, I tried to do the same for https://github.com/w3c/web-platform-tests/pull/6264 - also did not work. :(
Who should be looking into it?

zcorpan commented:
Filed https://github.com/w3c/web-platform-tests/issues/6293
